Fix Bug #5568, no dup notation warnings on repeated module imports

psteckler opened the pull request:
Uses an equality check to see if duplicate notation warning is needed. 

Added a test case for the reported bug.

The equality check uses (==) for the `not_interp` fields of the notation data, which might give bogus warnings. Using (=) for those fields often fails because of functional values (couldn't easily work out where those were). That's not ideal, of course.

This approach solves the reported bug, because we do get intensionally-equal fields in that case. In the test suite, there are no more warnings than there were in 8.6, where there was no equality check on added notations, so maybe this approach is acceptable.

JasonGross commented:
The test-suite file needs to be part of the `output/` tests; as it is right now, nothing is tested.

ejgallego commented:
Umm, this maybe deserves more general thinking about what to do with repeated warnings.

psteckler commented:
@JasonGross OK, I can make that change. For the other bugs in `bugs/closed`, what's being tested?

JasonGross commented:
What's being tested is that they compile correctly.  Generally the bug is "this code errors; it should pass", but that's not the case for this bug.

herbelin commented:
Hi, there is a test in `metasyntax.ml` to check if a notation already exists. So, especially if the warning is not consistent and depends of the exact form of the notation, I suspect that there is a bug in this function, namely `Notation.exists_notation_in_scope`. I would suggest to trace why `Notation.exists_notation_in_scope` does not do its job.

By the way, it may be the case that `Notation.exists_notation_in_scope` does not do its job in other situations. I see that it does not check the `only_print` flag, so, if a notation is reintroduced with only changing the `only_print` flag, it may fail to be updated. Indeed:
```coq
Module foo.
  Notation compose := (fun g f => (g,f)).
  Notation "g & f" := (compose g f) (at level 30).
End foo.
Module foo2.
  Notation compose := (fun g f => (g,f)).
  Notation "g & f" := (compose g f) (at level 30, only printing).
End foo2.
Import foo.
Import foo2.
Check (fun g f => (g,f)) 0 3. (* use the notation, is it expected? *)
```

psteckler commented:
@herbelin Thanks, I'm looking into that. I missed that check, because it's not called where the warning is issued.

JasonGross commented:
@herbelin 
> use the notation, is it expected?

Yes, I think it's expected.  New notations shadow old ones, they don't overwrite them, at least as far as printing is concerned.  I know of no way to remove a parsing notation, and I think this is a bug/missing feature, but I frequently make use of multiple `only printing` notations that turn multiple syntax trees into the same displayed expression.  This is why I think there should be a separate warning for shadowing printing notations, and shadowing parsing notations.

psteckler commented:
The failure of `exists_notation_in_scope` comes from the call to `interpretation_eq`, which calls `eq_notation_constr`, which recurses over a pair of terms, while keeping its first argument, a pair of lists  of variables, constant.

In Jason's example, the match case

``` NVar id1, NVar id2 -> Int.equal (List.index Id.equal id1 vars1) (List.index Id.equal id2 vars2) ```

raises an exception because the variables `id1` and `id2` don't appear in the lists `vars1` and `vars2`.

Maybe this code should check for list membership first, before checking for indexes.

I can make that change, and also add the printing-only equality check.

psteckler commented:
I changed the NVar, Nvar case in `eq_notation_constr` to use a helper that returns -1 if the variable doesn't exist in the list (maybe an option would be better). I also added an `onlyprinting` parameter to `exists_notation_in_scope`.

I put the bug code in `output/Notations.v` (at the top, the `Module` declaration wouldn't work at the end).

SkySkimmer commented:
Why `-1` instead of an option?

psteckler commented:
> Why -1 instead of an option?

The code is a little shorter, I can use Int.equal without pattern-matching on the option constructor.

SkySkimmer commented:
`Option.equal` lets you do that too.

psteckler commented:
Thanks, I didn't know about that one.

psteckler commented:
Uses an option now.

herbelin commented:
@JasonGross: OK

@psteckler: I think it is a bug if the variable does not occur. The `vars1` and `vars2` are supposed to form a closure over the variables of the notation.

psteckler commented:
@herbelin If we have two variables that are both not being closed over, should we considered them equal notation constrs? Or should we look for and fix an error in how the closures are being constructed, so that failure-to-close-over never occurs?

psteckler commented:
Here's what one of the interpretations looks like:

```
  ([(g, ((None, []), NtnTypeConstr)); (f, ((None, []), NtnTypeConstr))],
   NApp
    (NLambda (Names.Name.Name g0,
      NHole (Evar_kinds.BinderType (Names.Name.Name g0),
       Misctypes.IntroAnonymous, None),
      NLambda (Names.Name.Name f0,
       NHole (Evar_kinds.BinderType (Names.Name.Name f0),
        Misctypes.IntroAnonymous, None),
       NApp (NVar g0, [NVar f0]))),
    [NVar g; NVar f]))
```

psteckler commented:
The term that produces the `NApp` in the interpretation is `(fun g0 f0 => g0 f0) g f`. 

When `eq_notation_const` recurses into the `NLambda`, should it augment the list of variables with the `fun` formals, because those may bind variables in the `NLambda` body?

herbelin commented:
OK, I see, I forgot to take into account in `eq_notation_constr` those bound variables which are not bound in the notation, so that the `List.index` may fail brutally.

There are three kinds of variables in the rhs of a notation:
1. those free in the rhs and bound in the lhs (e.g. `Notation "# x" := (S x).`)
2. those binding in the rhs and bound in the lhs (e.g. `Notation "! x , P" := (forall x, P).` or `Notation ! x , P" := (forall x, x=0 -> P).`)
3. those binding in the rhs and not bound in the lhs (e.g. `Notation "ID" := (fun x => x).`)

The problem here is with variables of the 3rd kind. We should either consider them to be equal when they have the same name, or, if ever we also want to consider a notation such as `fun x => x` to be equal to a notation `fun y => y`, consider them when they are the same up to alpha-conversion.

So, a bit of work in all cases.

psteckler commented:
So the simple solution would be, if List.index fails for both variables, just check if they're the same name.

I'd say that handling the alpha-convertible case isn't so important.

psteckler commented:
I added the same-name check.

herbelin commented:
Perfect.

psteckler commented:
@herbelin did you want to approve the PR, then?
